lib/rom/plugins/relation/sql/auto_restrictions.rb

```diff
@@ -26,22 +23,31 @@ module SQL
         #
         # @api public
         module AutoRestrictions
-          EmptySchemaError = Class.new(ArgumentError) do
-            def initialize(klass)
-              super("#{klass} relation has no schema. " \
-                    "Make sure :auto_restrictions is enabled after defining a schema")
-            end
-          end
-
-          def self.included(klass)
-            super
+          def self.apply(klass)
             schema = klass.schema
             raise EmptySchemaError, klass if schema.empty?
+
             methods, mod = restriction_methods(schema)
             klass.include(mod)
             methods.each { |meth| klass.auto_curry(meth) }
           end
 
+          module FinalizationHook
+            def set_schema!(*)
+              super.tap { AutoRestrictions.apply(self) }
+            end
+          end
```

Comment:
I will try to get rid of this in favor of setup/finalization events like we talked about. It's fine for now though, it shouldn't block further development, so I'm merging this regardless.